### DISCUSSION TITLE
Admin-UI: Custom Configuration Tab in Realm Settings

### DISCUSSION THREAD

**Discussion by antikalk** [keycloak/keycloak#23486] [Ideas]
Coming back from https://github.com/keycloak/keycloak/issues/17732

At the moment it is not possible to edit realm attributes (for example custom realm attributes that are used in extensions) in the admin ui realm settings. Adding additional inputs for custom attributes would require to fork the new admin ui and therefor add additional workflows for keeping it in sync with the latest keycloak changes. We use additional realm attributes to configure our custom extensions. 

As some valid feedback was provided as to why https://github.com/keycloak/keycloak-ui/pull/4046 was not merged, I would like to propose a different solution: Provide an additional (conditionally visible) tab "Custom Configuration".

- Add realm attribute "allowCustomConfiguration" (default false)
- Add realm parameter "customConfiguration" (Map<String, String> or simple json string)
- Provide toggle under realm settings to allow for custom configuration (toggles realm attribute allowCustomConfiguration true/false)
- If allowCustomConfiguration is true, display tab "Custom Configuration"
- The "Custom Configuration" tab is build similiar to the user/groups attributes tab but does not render the realm attributes but the values from the realm parameter "customConfiguration" -> configurations can be added/edited/deleted  --- alternatively a simple json editor would possible be enough
- When a realm is saved and the "allowCustomConfiguration" is false, make sure to clear all custom configuration settings

Any thoughts?

**Comment by jonkoops** [comment]
I think ideally we'd like to have the ability for extensions to extend the realm settings with their own custom fields. In general we need a more flexible programmable API that allows extension developers to add menu items, tabs, sections, fields etc. to the UI without having to touch a line of JavaScript.

Internally we've been debating what would be a good approach for this, and we'll likely end up implementing a lot of our own UI in a new API that will allow us to do this as aforementioned. But we have yet to come up with a proposal for this as we're currently quite busy implementing the PatternFly 5 upgrade.

@ssilvert @agagancarczyk @edewit

**Reply by danielFesenmeyer** [reply]
We are also using custom realm attributes. It would be great if we could access these attributes via the UI.

**Reply by jonkoops** [reply]
> We are also using custom realm attributes. It would be great if we could access these attributes via the UI.

I think that is something we could add in a similar manner as we do for user attributes. WDTY @ssilvert?

**Reply by antikalk** [reply]
> > We are also using custom realm attributes. It would be great if we could access these attributes via the UI.
> 
> I think that is something we could add in a similar manner as we do for user attributes. WDTY @ssilvert?

See the PR I linked for the drawbacks of doing it that way (https://github.com/keycloak/keycloak-ui/pull/4046). Keycloak currently uses realm attributes for different realm settings, that maybe already have their corresponding "native" input to configure them. So simply rendering an Attribute-tab for managing all the realm attributes is not an option. At least the keycloak internal attributes would have to be filtered out...

**Reply by jonkoops** [reply]
Ah, I missed that discussion. Thanks that @antikalk.

**Reply by jbman** [reply]
A realm attribute view could only list realm attributes which are configured to be shown. The view would provide a single realm attribute to control this list. You add keys there to show those attributes.

However the main issue is that a single realm attribute can't be updated. The whole realm must be written, which could lead to lost updates if e.g. an other tool or extension writes at same time as an admin user.
So we would need a resource to update a single attribute together with a realm attribute view. 

I think this would be a good option, unless there is a decision for an other realm level configuration storage. Even extensions which are built on top of SPIs which support custom configuration (like authenticators) sometimes want to use realm level settings and not individual settings.
